packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -928,25 +1004,88 @@ class TextPainter {
   /// The [text] and [textDirection] properties must be non-null before this is
   /// called.
   void layout({ double minWidth = 0.0, double maxWidth = double.infinity }) {
-    assert(text != null, 'TextPainter.text must be set to a non-null value before using the TextPainter.');
-    assert(textDirection != null, 'TextPainter.textDirection must be set to a non-null value before using the TextPainter.');
-    // Return early if the current layout information is not outdated, even if
-    // _needsPaint is true (in which case _paragraph will be rebuilt in paint).
-    if (_paragraph != null && minWidth == _lastMinWidth && maxWidth == _lastMaxWidth) {
-      return;
+    assert(!maxWidth.isNaN);
+    assert(!minWidth.isNaN);
+    final InlineSpan? text = this.text;
+    if (text == null) {
+      throw StateError('TextPainter.text must be set to a non-null value before using the TextPainter.');
+    }
+    final TextDirection? textDirection = this.textDirection;
+    if (textDirection == null) {
+      throw StateError('TextPainter.textDirection must be set to a non-null value before using the TextPainter.');
+    }
+    final _TextPainterLayoutCache? cachedLayout = _layoutCache;
+    final double paintOffsetAlignment = _computePaintOffsetFraction(textAlign, textDirection);
+
+    // The assumption here is that if a Paragraph's width is already >= its
+    // maxIntrinsicWidth, further increasing the input width does not change its
+    // layout (but may change the paint offset if it's not left-aligned). This is
+    // true even for TextAlign.justify, since eot is also considered a hard
+    // line break so it will behave exactly the same as TextAlign.start when its
```

Comment:
I would probably make it clearer that this is because if the width exceeds the intrinsic width, by definition even justified text would not have wrapped. I had to read this several times to get the nuance here.